File: python/hashtable/leetcode/valid_anagram.py

# 有效的字母异位词

class Solution:

    # 排序
    def isAnagram_1(self, s: str, t: str) -> bool:
        return sorted(s) == sorted(t)

    # 使用一个哈希表
    def isAnagram_2(self, s: str, t: str) -> bool:
        if len(s) > len(t):
            return self.isAnagram_2(t, s)
        m = {}
        for c in s:
            m[c] = m[c] + 1 if c in m else 1
        for c in t:
            if c not in m or m[c] < 1:
                return False
            m[c] -= 1
        return True

File: java/algorithm/sort/leetcode/ValidAnagram.java
package algorithm.sort.leetcode;

import java.util.Arrays;

/**
 * @author roseduan
 * @time 2020/11/7 9:04 上午
 * @description 有效的字母异位词
 */
public class ValidAnagram {

    /**
     * 排序
     */
    public boolean isAnagram1(String s, String t) {
        char[] sCh = s.toCharArray();
        Arrays.sort(sCh);
        char[] tCh = t.toCharArray();
        Arrays.sort(tCh);
        return String.valueOf(sCh).equals(String.valueOf(tCh));
    }

    /**
     * 哈希表
     */
    public boolean isAnagram2(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }

        int[] data = new int[26];
        for (char c : s.toCharArray()) {
            data[c - 'a'] += 1;
        }
        for (char c : t.toCharArray()) {
            if (--data[c - 'a'] < 0) {
                return false;
            }
        }
        return true;
    }
}
